pagerduty/resource_pagerduty_user_notification_rule.go

```diff
@@ -42,46 +44,42 @@ func resourcePagerDutyUserNotificationRule() *schema.Resource {
 			"contact_method": {
 				Required: true,
 				Type:     schema.TypeMap,
-				Elem: &schema.Resource{
-					Schema: map[string]*schema.Schema{
-						"id": {
-							Type:     schema.TypeString,
-							Required: true,
-						},
-						"type": {
-							Type:     schema.TypeString,
-							Required: true,
-							ValidateFunc: validateValueFunc([]string{
-								"email_contact_method",
-								"phone_contact_method",
-								"push_notification_contact_method",
-								"sms_contact_method",
-							}),
-						},
-					},
+				// Using the `Elem` block to define specific keys for the map is currently not possible.
+				// The workaround described in SDK documentation is to confirm the required keys are set when expanding the Map object inside the resource code.
+				// See https://www.terraform.io/docs/extend/schemas/schema-types.html#typemap
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
 				},
+				ValidateDiagFunc: validation.MapKeyMatch(regexp.MustCompile("(id|type)"), "`contact_method` must only have `id` and `types` attributes"),
```

Comment:
An alternative that would have avoided to deport the validation logic is to convert the `contact_method` attribute to a `TypeList` or `TypeSet` with both `MinItems` and `MaxItems` set to 1, but then this would have required configuration changes for existing users to replace `contact_method = {}` assignments by `contact_method {}` blocks (without the equal sign).